### PR TITLE
Expose from_pretrained keyword arguments

### PR DESCRIPTION
**Pull request by mulhod**
Addresses #4650 

I'd like to get feedback about whether or not this change is reasonable/something you would consider merging. It is not finished yet. I did add a simple test just to demonstrate how it would work. I also tested `Predictor.from_path` locally with a model of my own based on `"bert-base-uncased"`.

### CODE REVIEW

File: allennlp/common/cached_transformers.py [hunk 1]
```diff
@@ -21,6 +21,7 @@ def get(
     make_copy: bool,
     override_weights_file: Optional[str] = None,
     override_weights_strip_prefix: Optional[str] = None,
+    from_pretrained_kwargs: Optional[Dict[str, Any]] = None,
```
Comment (epwalsh) [inline allennlp/common/cached_transformers.py:24]: I think we should just call this `**kwargs`.

**Review by epwalsh — CHANGES_REQUESTED** [review]
Hey @mulhod, I think this would be good to have, but I don't think we should add this argument to `Model.load`, `load_archive`, or `Predictor.from_path` since this is something that is specific to transformers, and you could just use the `overrides` argument to these functions in order to utilize this feature.

**Comment by mulhod** [comment]
I have revised the way it was being done to use `overrides` (which wasn't in `Predictor.from_path` -- I added it in).

I fear that this is still too complicated, but it appears to work. While I don't have a test for `Predictor.from_path`, I have tried that locally:
```
from json import dumps
from pathlib import Path

import ets_nlp_allennlp_models  # Internal package
from allennlp.predictors import Predictor

kwargs = {
    "transformers_from_pretrained_kwargs":
        {
            "local_files_only": True,
            "cache_dir": "/my/custom/location/for/torch/transformers"
        }
}

archive_path = Path("ets-nlp-allennlp-models/test_fixtures/models/general_regressor/model.tar.gz")
predictor = Predictor.from_path(archive_path, overrides=dumps(kwargs))
```

When the cache dir I pass in is empty/non-existent and `local_files_only` is `True`, this leads to the expected error. When present, the local files are used in that directory as opposed to the default location for the cache.

I would be happy to work on some more tests for this.

**Comment by mulhod** [comment]
I see the build failed in relation a formatting issue (easy fix) and a documentation issue (I added a `Path` type hint to the archive file parameter since it actually accepts a `Path` argument or a `str` argument. In any case, that is not related to this PR and can be taken out. It seemed like a small improvement to make. In any case, the tests pass otherwise, so I will hold off on that.

File: allennlp/common/cached_transformers.py [hunk 2]
```diff
@@ -74,9 +75,16 @@ def strip_prefix(s):
                     )
                 override_weights = {strip_prefix(k): override_weights[k] for k in valid_keys}
 
-            transformer = AutoModel.from_pretrained(model_name, state_dict=override_weights)
+            transformer = AutoModel.from_pretrained(
+                model_name,
+                state_dict=override_weights,
+                **kwargs.get("transformers_from_pretrained_kwargs", {}),
```
Comment (epwalsh) [inline allennlp/common/cached_transformers.py:81]:
```suggestion
                **kwargs,
```

File: allennlp/common/cached_transformers.py [hunk 3]
```diff
@@ -74,9 +75,16 @@ def strip_prefix(s):
...
+            )
         else:
-            transformer = AutoModel.from_pretrained(model_name)
+            transformer = AutoModel.from_pretrained(
+                model_name,
+                **kwargs.get("transformers_from_pretrained_kwargs", {}),
```
Comment (epwalsh) [inline allennlp/common/cached_transformers.py:86]:
```suggestion
                **kwargs,
```

File: allennlp/common/from_params.py [hunk 4]
```diff
@@ -207,6 +207,10 @@ def create_kwargs(
 
     if accepts_kwargs:
         kwargs.update(params)
+        if "transformers_from_pretrained_kwargs" in extras:
+            kwargs["transformers_from_pretrained_kwargs"] = extras[
+                "transformers_from_pretrained_kwargs"
+            ]
```
Comment (epwalsh) [inline allennlp/common/from_params.py:213]: I don't think this is necessary

```suggestion
```

File: allennlp/models/archival.py [hunk 5]
```diff
@@ -139,7 +140,7 @@ def load_archive(
 
     # Parameters
 
-    archive_file : `str`
+    archive_file : `str`/`Path`
```
Comment (epwalsh) [inline allennlp/models/archival.py:143]:
```suggestion
    archive_file : `Union[str, Path]`
```

File: allennlp/models/model.py [hunk 6]
```diff
@@ -287,13 +287,22 @@ def _load(
         )
 
         model_params = config.get("model")
+        extras = {}
+        if "transformers_from_pretrained_kwargs" in config:
+            extras["transformers_from_pretrained_kwargs"] = config[
+                "transformers_from_pretrained_kwargs"
+            ].as_dict()
```
Comment (epwalsh) [inline allennlp/models/model.py:294]: Shouldn't need this

```suggestion
```

File: allennlp/models/model.py [hunk 7]
```diff
@@ -287,13 +287,22 @@ def _load(
...
         # embeddings from.  We're now _loading_ the model, so those embeddings will already be
         # stored in our weights.  We don't need any pretrained weight file anymore, and we don't
         # want the code to look for it, so we remove it from the parameters here.
         remove_pretrained_embedding_params(model_params)
-        model = Model.from_params(vocab=vocab, params=model_params)
+        model = Model.from_params(
+            vocab=vocab,
+            params=model_params,
+            **extras,
+        )
```
Comment (epwalsh) [inline allennlp/models/model.py:305]: Shouldn't need this either

```suggestion
        model = Model.from_params(vocab=vocab, params=model_params)
```

File: allennlp/predictors/predictor.py [hunk 8]
```diff
@@ -247,7 +249,7 @@ def from_path(
 
         # Parameters
 
-        archive_path : `str`
+        archive_path : `str`/`Path`
```
Comment (epwalsh) [inline allennlp/predictors/predictor.py:252]:
```suggestion
        archive_path : `Union[str, Path]`
```

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 9]
```diff
@@ -57,12 +57,17 @@ def __init__(
         override_weights_file: Optional[str] = None,
         override_weights_strip_prefix: Optional[str] = None,
         gradient_checkpointing: Optional[bool] = None,
+        **kwargs,
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:60]:
```suggestion
        tokenizer_kwargs: Optional[Dict[str, Any]] = None,
        transformer_kwargs: Optional[Dict[str, Any]] = None,
```

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 10]
```diff
@@ -57,12 +57,17 @@ def __init__(
...
         super().__init__()
         from allennlp.common import cached_transformers
 
         self.transformer_model = cached_transformers.get(
-            model_name, True, override_weights_file, override_weights_strip_prefix
+            model_name,
+            True,
+            override_weights_file=override_weights_file,
+            override_weights_strip_prefix=override_weights_strip_prefix,
+            **kwargs,
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:70]:
```suggestion
            **(transformer_kwargs or {}),
```

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 11]
```diff
@@ -83,7 +88,10 @@ def __init__(
             self._scalar_mix = ScalarMix(self.config.num_hidden_layers)
             self.config.output_hidden_states = True
 
-        tokenizer = PretrainedTransformerTokenizer(model_name)
+        tokenizer = PretrainedTransformerTokenizer(
+            model_name,
+            tokenizer_kwargs=kwargs.get("transformers_from_pretrained_kwargs", {}),
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:93]:
```suggestion
            tokenizer_kwargs=tokenizer_kwargs,
```

**Review by epwalsh — CHANGES_REQUESTED** [review]
Hey @mulhod, sorry my initial comments were confusing. What I meant was that `cached_transformers.get()` should directly pass `**kwargs` to `AutoModel.from_pretrained()`, just like how `cached_transformers.get_tokenizer()` does.

I left some other suggestions for making the rest of the PR work with this.

**Comment by mulhod** [comment]
Thanks for the help @epwalsh. I applied your suggestions locally first to do some testing. While the tests I added pass (at least they did locally, not sure what the failure is yet), the arguments I'm passing through via `overrides` in `Predictor.from_path` are still not actually being passed through to `from_pretrained`. I set a breakpoint before all `from_pretrained` calls so that I can inspect the arguments and I don't see those I specified via `overrides`. I'm not sure what the issue is, but will look into it some more.

**Comment by epwalsh** [comment]
@mulhod can you show me what you're passing to `overrides`?

File: CHANGELOG.md [hunk 12]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   by adding a class-level variable called `authorized_missing_keys` to any PyTorch module that a `Model` uses.
   If defined, `authorized_missing_keys` should be a list of regex string patterns.
 - Added `FBetaMultiLabelMeasure`, a multi-label Fbeta metric. This is a subclass of the existing `FBetaMeasure`.
+- Added ability to pass `transformers`-related keyword arguments through to `from_pretrained` via `overrides`.
```
Comment (epwalsh) [inline CHANGELOG.md:20]:
```suggestion
- Added ability to pass additional key word arguments to `cached_transformers.get()`, which will be passed on to `AutoModel.from_pretrained()`.
- Added an `overrides` argument to `Predictor.from_path()`.
```

**Comment by mulhod** [comment]
@epwalsh I have made some more changes and was able to get it to work locally. I was using `overrides` incorrectly for one thing and that was my main problem. But I made some changes to my model code and some to this PR and was able to get the behavior I was interested in. This is the little test script that I was using:
```
from json import dumps
from pathlib import Path

import ets_nlp_allennlp_models  # internal ETS package
from allennlp.predictors import Predictor

kwargs = {
    "model.text_field_embedder.token_embedders.tokens": {
        "tokenizer_kwargs": {
            "local_files_only": True,
            "cache_dir": "/Users/mmulholland/.cache/torch/transformers"
        },
        "transformer_kwargs": {
            "local_files_only": True,
            "cache_dir": "/Users/mmulholland/.cache/torch/transformers"
        }
    },
    "model.seq2vec_encoder": {
        "transformer_kwargs": {
            "local_files_only": True,
            "cache_dir": "/Users/mmulholland/.cache/torch/transformers"
        }
    },
    "dataset_reader": {
        "tokenizer_kwargs": {
            "local_files_only": True,
            "cache_dir": "/Users/mmulholland/.cache/torch/transformers"
        },
        "transformer_kwargs": {
            "local_files_only": True,
            "cache_dir": "/Users/mmulholland/.cache/torch/transformers"
        }
    }
}

archive_path = Path("ets-nlp-allennlp-models/test_fixtures/models/general_regressor/model.tar.gz")
predictor = Predictor.from_path(archive_path, overrides=dumps(kwargs))
```

When `/Users/mmulholland/.cache/torch/transformers` has the cache files I need, the model loads correctly. When I move that directory to `/Users/mmulholland/.cache/torch/transformers.bak` and run the script again, it then fails to find the cache files, as expected.

File: allennlp/modules/seq2vec_encoders/bert_pooler.py [hunk 13]
```diff
@@ -31,7 +31,11 @@ class BertPooler(Seq2VecEncoder):
         Otherwise they will not.
     dropout : `float`, optional, (default = `0.0`)
         Amount of dropout to apply after pooling
-    """
+    transformer_kwargs: `Dict[str, Any]`, optional (default = `None`)
+        Dictionary with
+        [additional arguments](https://github.com/huggingface/transformers/blob/155c782a2ccd103cf63ad48a2becd7c76a7d2115/transformers/modeling_utils.py#L253)
+        for `AutoTokenizer.from_pretrained`.
```
Comment (mulhod) [inline allennlp/modules/seq2vec_encoders/bert_pooler.py:37]:
```suggestion
        for `AutoModel.from_pretrained`.
```

**Review by epwalsh — CHANGES_REQUESTED** [review]
@mulhod thank you, this looks really good!

I just had a few minor comments. If you can fix those up we'll get this merged 🎉 

As a side note, I feel like the `overrides` argument throughout the library should accept a JSON string or a Python dictionary. That would make it much easier to use from Python directly. What do you think?

**Comment by mulhod** [comment]
Thanks for the suggestions. I'll take care of the tests stuff.

> As a side note, I feel like the `overrides` argument throughout the library should accept a JSON string or a Python dictionary. That would make it much easier to use from Python directly. What do you think?

Yes, this would be great and probably a pretty easy change. Should it be taken care of here?

**Comment by epwalsh** [comment]
> Yes, this would be great and probably a pretty easy change. Should it be taken care of here?

I think that would be best as a separate PR. Feel free to assign me as a reviewer if you want to take that on.

**Review by epwalsh — APPROVED** [review]
LGTM!

**Pull request merged**